sdk/cosmos/src/clients/cosmos_client.rs

```diff
@@ -177,23 +177,8 @@ impl CosmosClient {
     }
 
     /// Create a database
-    pub async fn create_database<S: AsRef<str>>(
-        &self,
-        ctx: Context,
-        database_name: S,
-        options: CreateDatabaseOptions,
-    ) -> crate::Result<CreateDatabaseResponse> {
-        let mut request = self.prepare_request_pipeline("dbs", http::Method::POST);
-
-        let mut pipeline_context = PipelineContext::new(ctx, ResourceType::Databases.into());
-
-        options.decorate_request(&mut request, database_name.as_ref())?;
-        let response = self
-            .pipeline()
-            .send(&mut pipeline_context, &mut request)
-            .await?;
-
-        Ok(CreateDatabaseResponse::try_from(response).await?)
+    pub fn create_database<S: AsRef<str>>(&self, database_name: S) -> CreateDatabaseBuilder {
+        CreateDatabaseBuilder::new(self.clone(), database_name.as_ref().to_owned())
```

Comment:
I'm trying to figure out the best way to pass parameters in #520 #527 and `as_ref().to_owned()` may [be an anti-pattern](https://www.philipdaniels.com/blog/2019/rust-api-design/). Should `impl Into<String>` be used instead. Could the Builders avoid some cloning with `Cow`?

Reply:
> Should `impl Into<String>` be used instead.

I think your suggestion to use `impl Into<String>` is a good one; it evades an unnecessary allocation when an owned string is passed directly. This should make for an easy improvement.

---

>  Could the Builders avoid some cloning with `Cow`?

Potentially we could; for example using `impl Into<Cow<'static, str>>` as the API bound could potentially save us from having to perform an allocation when passing string slices as well. But I think if we go down this path we should have guidelines in place. As we know from experience it's really easy to accidentally push lifetimes into customer-facing APIs, which ends up having real consequences for user experience. And `Cow` can make it easy to fall into that trap.

In general I'd prefer it if we took a gradual approach to this. The change to using `Into<String>` over `AsRef<str>` should be an easy one we can do straight away. And once we're comfortable with the general shape of the APIs, we can look at optimizing things further by seeing if we can make the switch over to the slightly more complicated `Into<Cow<'static, str>>`. How does that sound?